packages/fxa-auth-server/lib/routes/subscriptions/paypal.ts

```diff
@@ -135,6 +114,137 @@ export class PayPalHandler extends StripeHandler {
       subscription: filterSubscription(subscription),
     };
   }
+
+  /**
+   * Updates the billing agreement for a user with a new PayPal token.
+   *
+   * Updates are not allowed if there's:
+   *  - No paypal sourced subscriptions active/past_due
+   *  - An existing billing agreement on file.
+   *
+   * If there's open invoices, attempts to pay them will be made in
+   * the background.
+   */
+  async updatePaypalBillingAgreement(request: AuthRequest) {
+    this.log.begin('subscriptions.createSubscriptionWithPaypal', request);
+    const { uid, email } = await handleAuth(this.db, request.auth, true);
+    await this.customs.check(request, email, 'createSubscriptionWithPaypal');
+
+    let customer = await this.stripeHelper.customer({
+      uid,
+      email,
+    });
+
+    if (!customer) {
+      throw error.unknownCustomer(uid);
+    }
+
+    if (this.stripeHelper.getCustomerPaypalAgreement(customer)) {
+      throw error.billingAgreementExists(customer.id);
+    }
+
+    const paypalSubscription = customer.subscriptions?.data.find(
+      (sub) =>
+        ['active', 'past_due'].includes(sub.status) &&
+        sub.collection_method === 'send_invoice'
+    );
+    if (!paypalSubscription || !customer.currency) {
+      throw error.internalValidationError('updatePaypalBillingAgreement', {
+        customerId: customer.id,
+        message: 'Cannot add PayPal with no paypal subscriptions.',
+        uid,
+      });
```

Comment:
If the error is thrown when the customer doesn't have a currency, then this message is misleading.